transmart-core-db/grails-app/services/org/transmartproject/db/userqueries/UserQuerySetService.groovy

```diff
@@ -0,0 +1,267 @@
+import org.transmartproject.db.querytool.QuerySet
+import org.transmartproject.db.querytool.QuerySetDiff
+import org.transmartproject.db.querytool.QuerySetInstance
+import org.transmartproject.db.user.User as DbUser
+import org.transmartproject.core.users.UsersResource
+import org.transmartproject.db.accesscontrol.AccessControlChecks
+
+import static org.transmartproject.db.multidimquery.DimensionImpl.PATIENT
+
+@Transactional
+class UserQuerySetService implements UserQuerySetResource {
+
+    @Autowired
+    UsersResource usersResource
+
+    @Autowired
+    UserQueryService userQueryService
+
+    @Autowired
+    MultidimensionalDataResourceService multiDimService
+
+    @Autowired
+    AccessControlChecks accessControlChecks
+
+    SessionFactory sessionFactory
+
+    @Override
+    Integer scan(User currentUser) {
+        int numberOfResults = 0
+        DbUser user = (DbUser) usersResource.getUserFromUsername(currentUser.username)
+        if (!user.admin) {
+            throw new AccessDeniedException('Only allowed for administrators.')
+        }
+        // get list of all not deleted queries per user
+        List<UserQuery> userQueries = userQueryService.listSubscribed()
+        if (!userQueries) {
+            log.info "No subscribed queries were found."
+            return numberOfResults
+        }
+
+        for (query in userQueries) {
+
+            List<QuerySetInstance> previousQuerySetInstances = getInstancesForLatestQuerySet(query.id)
+            ArrayList<Long> newPatientIds = getPatientsForQuery(query, user)
+
+            if (createSetWithDiffEntries(previousQuerySetInstances*.objectId, newPatientIds, (Query) query)) {
+                numberOfResults++
+            }
+
+        }
+        return numberOfResults
+    }
+
+    @Override
+    List<UserQuerySetInstance> getSetInstancesByQueryId(Long queryId, User currentUser, int firstResult, Integer numResults) {
+        // check access and if query exists
+        userQueryService.get(queryId, currentUser)
+
+        def session = sessionFactory.currentSession
+        Criteria criteria = session.createCriteria(UserQuerySetInstance, "querySetInstances")
+                .createAlias("querySetInstances.querySet", "querySet", JoinType.INNER_JOIN)
+                .createAlias("querySet.query", "query", JoinType.INNER_JOIN)
+                .add(Restrictions.eq('query.id', queryId))
+                .add(Restrictions.eq('query.deleted', false))
+                .addOrder(Order.desc('querySet.createDate'))
+                .setResultTransformer(Criteria.ALIAS_TO_ENTITY_MAP)
+                .setFirstResult(firstResult)
+        if(numResults) {
+            criteria.setMaxResults(numResults)
+        }
+        def result = criteria.list()
+        if (!result) {
+            return []
+        }
+
+        DbUser user = (DbUser) usersResource.getUserFromUsername(currentUser.username)
+        if (!user.admin && result.query.first().username != currentUser.username) {
```

Comment:
This was already checked on line 1, right?

Reply:
I removed 1st line, because it was blocking an access to all querySets for admin users .

Reply:
I don't understand. Line 84 has `userQueryService.get(queryId, currentUser)`, but nothing happens with the result.
And why would an admin want to fetch another user's query set?

Reply:
Sorry, forgot to pull. I will look again.

Reply:
Question remains: why not just include a restriction on the username in the query? I don't see a use for admins to retrieve arbitrary sets.